Fix #2658, #299: Replace builder() with Factory in SingleTypeBuilder and MultiTypeBuilder

KevinGitonga opened the pull request:
Fixes #2658
Fixes #299

This PR includes a set of changes to the BindableAdapter including creating a Factory Instance of the SingleTypeAdapter and MultiTypeAdapter instead of the currently used Builder Pattern to enable it to be Injected into Fragments and ListView instances used throughout the App. 
 
See also https://github.com/KevinGitonga/oppia-android/pull/1 for an explanation on some sub-parts of the PR.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

KevinGitonga committed:
Replace builder() with Factory in SingleTypeBuilder and MultiTypeBuilder.

BenHenning commented on app/src/main/java/org/oppia/android/app/settings/profile/ProfileListFragmentPresenter.kt:
```diff
@@ -53,7 +53,7 @@ class ProfileListFragmentPresenter @Inject constructor(
   private fun createRecyclerViewAdapter(): BindableAdapter<Profile> {
     return BindableAdapter.SingleTypeBuilder
-      .newBuilder<Profile>()
+      .Factory(fragment).create<Profile>()
```
Here & elsewhere: Since ``Factory`` is injectable, we should be injecting it, not constructing it.

KevinGitonga replied:
Noted @BenHenning will make the changes, Will check why i missed this while refactoring and AS didn't show error while building the code.

BenHenning commented on app/src/main/java/org/oppia/android/app/recyclerview/BindableAdapter.kt:
```diff
@@ -95,63 +93,15 @@ class BindableAdapter<T : Any> internal constructor(
-  abstract class BaseBuilder<BuilderType : Any> {
-    /**
-     * A [WeakReference] to a [LifecycleOwner] for databinding inflation. See [setLifecycleOwner].
-     * Note that this needs to be a weak reference so that long-held references to the adapter do
```
This part still seems relevant.

I'm actually wondering if we should keep the ``BaseBuilder``, but just change it so that the lifecycle owner is passed in via the constructor rather than through ``setLifecycleOwner``.

KevinGitonga replied:
Yeah Ben i think we can keep it i noted it is holding a weak reference of the LifecycleOwner. As described to prevent memory leaks.

BenHenning commented on app/src/main/java/org/oppia/android/app/testing/DragDropTestActivityPresenter.kt:
```diff
@@ -24,7 +25,7 @@ class DragDropTestActivityPresenter @Inject constructor(private val activity: Ap
   private fun createBindableAdapter(): BindableAdapter<String> {
     return BindableAdapter.SingleTypeBuilder
-      .newBuilder<String>()
+      .Factory(Fragment()).create<String>()
```
This isn't ideal, I think. We should be using a real fragment. Perhaps this could be moved to a test fragment that's added to the test activity? I don't think we even want to encourage something like this since it's not going to behave correctly in cases when a real lifecycle owner is actually needed.

KevinGitonga replied:
Hi @BenHenning i did this intentionally to get the refactor working so that afterwards i can embed a fragment inside the Activity and inject the Adapter Factory.

BenHenning replied:
Given that the refactor seems to be solid now, let's get this comment resolved along with the others (so that everything is properly hooked up & injected per the new design).

BenHenning replied:
I think this is still unresolved.

KevinGitonga replied:
Updating to inflate fragment.

BenHenning reviewed (changes requested):
Thanks @KevinGitonga. At a high-level, I think we need to be injecting the factory since that's the intent of the issue being fixed.

For the PR itself:
- Please make sure that the title is complete (it shouldn't be abbreviated, and probably needs more clarity on what the PR is actually doing), and follows the format: ``Fix #<issuenumber>: Title``.
- Please make sure the PR description includes ``Fixes #<issuenumber>`` as part of the explanation. I also suggest rewording the explanation to be clearer on what's actually being solved, and why the solution makes sense.
- If the UI section doesn't need to be filled out, please include a brief justification for why it doesn't.
- It looks like a bunch of CI checks are failing. Once you address my comments, I suggest going through and making sure that those checks are passing before sending the PR back for review.

KevinGitonga committed:
Fix failing tests after refactor, Create Fragment for DragDropTestActivity.kt.

KevinGitonga committed:
Inject the BindableAdapter.kt factory into the fragment and listview instances.

KevinGitonga committed:
Inject the BindableAdapter.kt factory into the fragment and listview instances.

KevinGitonga committed:
Inject the BindableAdapter.kt factory into the fragment and listview instances.

KevinGitonga committed:
Inject the BindableAdapter.kt factory into the fragment and listview instances.

KevinGitonga committed:
Add 'androidx.work:work-testing:2.4.0' in androidTestImplementation to fix tests.

KevinGitonga committed:
Fix lint formatting issue in updated files.

KevinGitonga committed:
Fix some of the failing tests after fixing #2658.

KevinGitonga committed:
Fix some of the failing tests after fixing #2658.

KevinGitonga committed:
Fix more failing tests for #2658.

BenHenning commented:
@KevinGitonga given that this PR is still in draft, did you mean for me to perform another review pass? It's not clear what you're looking for here since you didn't leave an explicit ask message on the PR (or maybe I missed it?).

If the PR is ready for full review, you should click the 'Ready for review' button. If it's still in draft, please follow up with what you'd like me to specifically look at.

KevinGitonga committed:
Refactor StatePlayerRecyclerViewAssembler.kt to inject MultiTypeAdapterFactory

KevinGitonga committed:
Fix failing tests due to missed refactor of QuestionPlayerFragmentPresenter.kt to inject MultiTypeAdapter Factory.

BenHenning commented on app/build.gradle:
```diff
@@ -208,6 +208,7 @@ dependencies {
       'androidx.test.ext:junit:1.1.1',
       'com.github.bumptech.glide:mocks:4.11.0',
       'com.google.truth:truth:1.1.3',
+      'androidx.work:work-testing:2.4.0',
```
Wasn't this introduced in your other PR? I suggest syncing to the latest develop so that unrelated changes don't need to be in this branch.

KevinGitonga replied:
Yeah @BenHenning updated and synced branch to pick up latest develop.

BenHenning commented on app/src/main/java/org/oppia/android/app/administratorcontrols/AdministratorControlsFragmentPresenter.kt:
```diff
@@ -78,7 +78,7 @@ class AdministratorControlsFragmentPresenter @Inject constructor(
     BindableAdapter<AdministratorControlsItemViewModel> {
       return BindableAdapter.MultiTypeBuilder
-        .newBuilder<AdministratorControlsItemViewModel, ViewType> { viewModel ->
+        .Factory(fragment).create<AdministratorControlsItemViewModel, ViewType> { viewModel ->
```
This factory should be injected instead of constructed here.

KevinGitonga replied:
Yeah noted have noticed i still missed some classes during refactor. Will have to run global search to cover all required classes.

BenHenning replied:
FWIW changing the class visibilities as suggested below should trigger build failures for any case where this hasn't been corrected, so an audit may not be necessary (since the system can do it for you).

BenHenning commented on app/src/main/java/org/oppia/android/app/administratorcontrols/learneranalytics/ProfileAndDeviceIdFragmentPresenter.kt:
```diff
@@ -38,16 +40,15 @@ class ProfileAndDeviceIdFragmentPresenter @Inject constructor(
-          is SyncStatusItemViewModel -> ProfileListItemViewType.SYNC_STATUS
-          else -> error("Encountered unexpected view model: $viewModel")
-        }
+    return multiTypeBuilderFactory.create<ProfileListItemViewModel,
```
If we name the injected field as ``adapterFactory`` will this be able to fit on one line so that it doesn't need to be line-wrapped in this way?

KevinGitonga replied:
Yeah it should be able to.

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt:
```diff
@@ -94,14 +95,14 @@ class MarkChaptersCompletedFragmentPresenter @Inject constructor(
-          is ChapterSummaryViewModel -> ViewType.VIEW_TYPE_CHAPTER
-          else -> throw IllegalArgumentException("Encountered unexpected view model: $viewModel")
-        }
+    return multiTypeBuilderFactory.create<MarkChaptersCompletedItemViewModel,
```
Ditto here--let's see if we can wrap this a bit more cleanly.

KevinGitonga replied:
Changes applied.

BenHenning commented on app/src/main/java/org/oppia/android/app/home/promotedlist/ComingSoonTopicsListView.kt:
```diff
@@ -32,6 +32,12 @@ class ComingSoonTopicsListView @JvmOverloads constructor(
   @Inject
   lateinit var oppiaLogger: OppiaLogger
 
+  @Inject
+  lateinit var fragment: Fragment
```
Is this needed?

KevinGitonga replied:
Refactored out.

BenHenning commented on app/src/main/java/org/oppia/android/app/home/promotedlist/PromotedStoryListView.kt:
```diff
@@ -32,6 +32,12 @@ class PromotedStoryListView @JvmOverloads constructor(
   @Inject
   lateinit var oppiaLogger: OppiaLogger
 
+  @Inject
+  lateinit var fragment: Fragment
```
Ditto--is this needed?

KevinGitonga replied:
Not needed as of now, refactored out.

BenHenning replied:
It seems seems to be present--did you mean to remove it?

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/DragDropSortInteractionView.kt:
```diff
@@ -53,6 +53,12 @@ class DragDropSortInteractionView @JvmOverloads constructor(
   @Inject
   lateinit var viewBindingShim: ViewBindingShim
 
+  @Inject
+  lateinit var fragment: Fragment
```
Ditto.

KevinGitonga replied:
Applied changes.

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/SelectionInteractionView.kt:
```diff
@@ -45,6 +45,12 @@ class SelectionInteractionView @JvmOverloads constructor(
   @Inject
   lateinit var bindingInterface: ViewBindingShim
 
+  @Inject
+  lateinit var fragment: Fragment
```
Ditto.

KevinGitonga replied:
Applied changes.

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/SelectionInteractionView.kt:
```diff
@@ -108,7 +113,7 @@ class SelectionInteractionView @JvmOverloads constructor(
       SelectionItemInputType.RADIO_BUTTONS ->
         BindableAdapter.SingleTypeBuilder
-          .newBuilder<SelectionInteractionContentViewModel>()
+          .Factory(fragment).create<SelectionInteractionContentViewModel>()
```
We should use the injected factory here rather than reconstructing it.

KevinGitonga replied:
Applied changes.

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/StatePlayerRecyclerViewAssembler.kt:
```diff
@@ -1183,6 +1183,17 @@ class StatePlayerRecyclerViewAssembler private constructor(
+    /**
+     * Add AdapterBuilderFactory passed through injection from [StateFragmentPresenter]
+     */
+    fun addAdapterBuilderFactory
+    (multiTypeBuilderFactory: BindableAdapter.MultiTypeBuilder.Factory): Builder {
+      adapterBuilder = multiTypeBuilderFactory.create {
+        it.viewType
+      }
+      return this
+    }
```
Do we actually need this? I think you can inject the factory in the assembler's own injectable factory to avoid needing the extra plumbing.

KevinGitonga replied:
Have refactored this too

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/StatePlayerRecyclerViewAssembler.kt:
```diff
@@ -1117,7 +1117,7 @@ class StatePlayerRecyclerViewAssembler private constructor(
     ): BindableAdapter<StringList> {
       return BindableAdapter.SingleTypeBuilder
-        .newBuilder<StringList>()
+        .Factory(fragment).create<StringList>()
```
For this & the one below: we should be using an injected factory rather than constructing a new one.

KevinGitonga replied:
Patched.

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/ProfileChooserFragmentPresenter.kt:
```diff
@@ -148,10 +149,10 @@ class ProfileChooserFragmentPresenter @Inject constructor(
-      .newBuilder<ProfileChooserUiModel, ProfileChooserUiModel.ModelTypeCase>(
-        ProfileChooserUiModel::getModelTypeCase
-      )
+    return multiTypeAdapterBuilder.create<ProfileChooserUiModel,
```
Ditto for this variable name--can we line wrap this more cleanly?

KevinGitonga replied:
Patched.

BenHenning commented on app/src/main/java/org/oppia/android/app/profileprogress/ProfileProgressFragmentPresenter.kt:
```diff
@@ -26,6 +26,9 @@ class ProfileProgressFragmentPresenter @Inject constructor(
   lateinit var viewModel: ProfileProgressViewModel
 
+  @Inject
+  lateinit var multiTypeAdapterBuilder: BindableAdapter.MultiTypeBuilder.Factory
```
This (& technically the view model above) should be injected via the constructor, not field injection.

KevinGitonga replied:
Yeah updating to constructor injection.

BenHenning commented on app/src/main/java/org/oppia/android/app/recyclerview/BindableAdapter.kt:
```diff
@@ -242,10 +243,11 @@ class BindableAdapter<T : Any> internal constructor(
-      /** Returns a new [SingleTypeBuilder]. */
-      inline fun <reified T : Any> newBuilder(): SingleTypeBuilder<T> {
-        return SingleTypeBuilder(T::class)
+    class Factory @Inject constructor(
```
Please add KDocs for this & the next factor, and for their ``create`` methods.

KevinGitonga replied:
Added.

BenHenning commented on app/src/main/java/org/oppia/android/app/recyclerview/BindableAdapter.kt:
```diff
@@ -150,7 +150,8 @@ class BindableAdapter<T : Any> internal constructor(
   class SingleTypeBuilder<T : Any>(
-    private val dataClassType: KClass<T>
+    private val dataClassType: KClass<T>,
+    private val fragment: Fragment
```
Should we maybe pass this into ``BaseBuilder``'s constructor and remove ``setLifecycleOwner``? This allows us to leverage the primary benefit of being able to inject the fragment. Ditto for the multi-type below.

``BaseBuilder`` can also be private now, so we should change it to be private to minimize visibility.

Furthermore, both builder constructors can and should now be private since they're created via injectable factories.

KevinGitonga replied:
@BenHenning given this updated https://gist.github.com/KevinGitonga/4c6e7e457949a0c6abee2608b51c9a64 How do i refactor BaseBuilder class to still return BuilderType.

KevinGitonga replied:
@BenHenning to be specific, attaching the lifecycleowner from parent fragment if not null to the weak reference lifecycle holder since not all parent users of BindableAdapter need to attach lifecycle as noted previously.

BenHenning replied:
setLifecycleOwner shouldn't be needed anymore, which means BuilderType wouldn't be needed anymore.

One of the main goals of this refactor was to automatically attach the lifecycle owner in all cases since it's never wrong to attach, and it's easy to forget (which leads to hard-to-debug situations where LiveData simply doesn't work via databinding).

KevinGitonga replied:
Well noted Ben this makes sense.

BenHenning replied:
I think the only thing left for this comment is ensuring that the constructors become private for classes that don't need to be instantiated directly.

BenHenning replied:
Left a follow up comment regarding the constructor visibility.

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/lessons/TopicLessonsFragmentPresenter.kt:
```diff
@@ -208,7 +208,7 @@ class TopicLessonsFragmentPresenter @Inject constructor(
   private fun createChapterRecyclerViewAdapter(): BindableAdapter<ChapterSummaryViewModel> {
     return BindableAdapter.SingleTypeBuilder
-      .newBuilder<ChapterSummaryViewModel>()
+      .Factory(fragment).create<ChapterSummaryViewModel>()
```
Should be injected.

KevinGitonga replied:
Updated.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/recyclerview/BindableAdapterTest.kt:
```diff
@@ -163,8 +163,10 @@ class BindableAdapterTest {
+    val testFragment = Fragment()
     // Set up the adapter to be used for this test.
-    TestModule.testAdapterFactory = { createSingleViewTypeNoDataBindingBindableAdapter() }
+    TestModule.testAdapterFactory =
```
This mechanism can be simplified a bit, I think. We should update all of the ``create*Adapter(`` helpers to take in their respective factory, then we can update: https://github.com/oppia/oppia-android/blob/78f4f5255ded9c4bd64b582e0c50c5dd38a923e5/app/src/main/java/org/oppia/android/app/testing/BindableAdapterTestFragmentPresenter.kt#L35

to pass in both factories, then we can select whichever is needed & pass it to the corresponding helper when setting this factory, e.g.:

```kotlin
TestModule.testAdapterFactory = { singleFactory, _ -> createSingleViewTypeNoDataBindingBindableAdapter(singleFactory) }
```

This will ensure the correct fragment is hooked up since the test fragment will be injecting the factories directly. It also ensures that we're actually injecting the factories rather than constructing them like below.

KevinGitonga replied:
@BenHenning applied the refactor however classes which were testing lifecycle implementations have been affected and failing e.g testSingleTypeAdapter_setLifecycleOwnerTwice_throwsException. Please advice since we are not manually attaching lifecycle to the BindableAdapter and it is done through injection?

BenHenning replied:
Anything that is testing setLifecycleOwner behaviors can be removed. We should still verify that the lifecycle binding works, but since we can't change it now any tests that require setLifecycleOwner to be called after the initial binding, or not at all, can be removed as they're no longer possible behaviors.

KevinGitonga replied:
Okay Ben removing all tests covering LifecycleOwner

BenHenning reviewed (changes requested):
Thanks @KevinGitonga! I think this PR is definitely on the right track, but I noticed a bunch of places that will need some changes. PTAL.

KevinGitonga committed:
Merge branch 'oppia:develop' into enable_dagger_graph_access_for_bindable_adapter

KevinGitonga committed:
More updates as part of refactoring for issue #2658.

KevinGitonga committed:
Refactor PromotedStoryListView.kt BindableAdapterTest.kt classes to provide Adapter through injection for isuue #2658.

BenHenning commented on app/src/main/java/org/oppia/android/app/recyclerview/BindableAdapter.kt:
```diff
@@ -99,34 +99,13 @@ class BindableAdapter<T : Any> internal constructor(
-      // the builder & pass their own type in.
-      @Suppress("UNCHECKED_CAST") return this as BuilderType
-    }
+    private var lifecycleOwnerRef: WeakReference<LifecycleOwner> = WeakReference(fragment)
```
Can be a val, I think, since it'll never change.

KevinGitonga replied:
Yeah for sure

BenHenning commented on app/src/main/java/org/oppia/android/app/recyclerview/BindableAdapter.kt:
```diff
@@ -99,34 +99,13 @@ class BindableAdapter<T : Any> internal constructor(
    * [SingleTypeBuilder] or [MultiTypeBuilder] instead.
    */
-  abstract class BaseBuilder<BuilderType : Any> {
+  abstract class BaseBuilder(fragment: Fragment) {
```
Can we make this constructor private now?

KevinGitonga replied:
Patch applied

BenHenning replied:
I think you might've made ``fragment`` a private val rather than making the constructor private. You should instead do:

```kotlin
abstract class BaseBuilder private constructor(fragment: Fragment) {
```

BenHenning commented on app/src/main/java/org/oppia/android/app/recyclerview/BindableAdapter.kt:
```diff
@@ -223,10 +202,11 @@ class BindableAdapter<T : Any> internal constructor(
         object : BindableViewHolder<T>(binding.root) {
           override fun bind(data: T) {
             setViewModel(binding, data)
+            binding.lifecycleOwner = getLifecycleOwner()
```
Can you please add a comment here explaining why this has to happen here specifically (per the issues that you ran into)?

KevinGitonga replied:
Yeah sure thing.

BenHenning commented on app/src/main/java/org/oppia/android/app/recyclerview/BindableAdapter.kt:
```diff
@@ -374,9 +359,14 @@ class BindableAdapter<T : Any> internal constructor(
       )
     }
 
+    /** Fragment injectable factory to create new [MultiTypeBuilder] */
```
```suggestion
    /** Fragment injectable factory to create new [MultiTypeBuilder]. */
```

KevinGitonga replied:
Noted.

BenHenning commented on app/src/main/java/org/oppia/android/app/recyclerview/BindableAdapter.kt:
```diff
@@ -243,9 +223,13 @@ class BindableAdapter<T : Any> internal constructor(
     class Factory @Inject constructor(
       val fragment: Fragment
     ) {
+      /**
+       * Returns a new [SingleTypeBuilder] for the specified Data class type.
+       */
```
```suggestion
      /** Returns a new [SingleTypeBuilder] for the specified Data class type. */
```

Can fit on one line. Ditto below.

KevinGitonga replied:
Noted.

BenHenning commented on app/src/main/java/org/oppia/android/app/home/promotedlist/PromotedStoryListView.kt:
```diff
@@ -32,19 +32,51 @@ class PromotedStoryListView @JvmOverloads constructor(
+
   override fun onAttachedToWindow() {
-    super.onAttachedToWindow()
+    try {
```
I know you were using this try-catch to test, but we shouldn't actually keep it in (we should fix the issue, instead). Can it be removed now?

KevinGitonga replied:
Okay Noted.

BenHenning commented on app/src/main/java/org/oppia/android/app/home/promotedlist/PromotedStoryListView.kt:
```diff
@@ -32,19 +32,51 @@ class PromotedStoryListView @JvmOverloads constructor(
-    val snapHelper = StartSnapHelper()
-    onFlingListener = null
-    snapHelper.attachToRecyclerView(this)
+  private fun checkIfComponentsInitialized() {
```
Or perhaps: ``maybeInitializeAdapter``?

KevinGitonga replied:
Noted sounds well too.

BenHenning commented on app/src/main/java/org/oppia/android/app/home/promotedlist/PromotedStoryListView.kt:
```diff
@@ -61,15 +100,15 @@ class PromotedStoryListView @JvmOverloads constructor(
     }
-    if (newDataList == null) {
-      oppiaLogger.w(PROMOTED_STORY_LIST_VIEW_TAG, "Failed to resolve new story list data")
+    if (::promotedDataList.isInitialized) {
```
Isn't this guaranteed now since the method is only called after everything is initialized? It seems like we can simplify this code, then, as the error case shouldn't happen anymore.

KevinGitonga replied:
Yeah Ben it is guaranteed now. Updating.

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/StatePlayerRecyclerViewAssembler.kt:
```diff
@@ -887,11 +887,13 @@ class StatePlayerRecyclerViewAssembler private constructor(
-      StateItemViewModel::viewType
-    )
+
+    private var adapterBuilder: BindableAdapter.MultiTypeBuilder<StateItemViewModel,
```
Can this just be a ``val``? It doesn't seem to need to change.

KevinGitonga replied:
Noted.

BenHenning reviewed (commented):
Apologies for the delay.

Thanks @KevinGitonga! Just had a few more comments. I think that once the remaining comments are addressed & CI is passing that this PR will be more or less completed, or very close. PTAL at the latest comments & CI failures.

KevinGitonga committed:
Fix issues as advised by code review.

KevinGitonga committed:
Fix issue causing some of failing tests.

KevinGitonga committed:
Fix issue causing some of failing tests.

KevinGitonga committed:
Revert unused resources, to fix Static check test failure.

KevinGitonga committed:
Add KDocs to implemented public functions, to fix KDocs Static check test failure.

KevinGitonga committed:
Add exemption for test to DragDropTestFragmentPresenter and DragDropTestFragment since they are actual test classes.

BenHenning commented on app/src/main/java/org/oppia/android/app/recyclerview/BindableAdapter.kt:
```diff
@@ -99,13 +99,13 @@ class BindableAdapter<T : Any> internal constructor(
    * [SingleTypeBuilder] or [MultiTypeBuilder] instead.
    */
-  abstract class BaseBuilder(fragment: Fragment) {
+  abstract class BaseBuilder(private val fragment: Fragment) {
```
We should probably keep the property as an argument and not save it as a class member (otherwise we're leaking the context and the weak reference becomes ineffective).

KevinGitonga replied:
Thanks @BenHenning taking a look at the nits.

BenHenning commented on app/src/main/java/org/oppia/android/app/recyclerview/BindableAdapter.kt:
```diff
@@ -206,6 +206,8 @@ class BindableAdapter<T : Any> internal constructor(
           override fun bind(data: T) {
             setViewModel(binding, data)
+            // Attach lifecycleOwner after viewModel has been attached to the view.
+            // Attaching lifecycleOwner before will cause a null pointer exception crash on [HomeFragment]
```
For here & the version below, two things:
1. I don't think we can actually link to HomeFragment from here since this adapter doesn't import it (plus, markdown links aren't meant to work in single-line comments unlike KDocs).
2. Instead of talking about HomeFragment, could these comments explain why the NPE can sometimes happen? E.g. something like: "Attaching the lifecycle owner before view model initialization can sometimes cause a NullPointerException because X."

KevinGitonga replied:
Makes sense Ben have updated comments.

BenHenning commented on app/src/main/java/org/oppia/android/app/testing/DragDropTestActivityPresenter.kt:
```diff
@@ -1,61 +1,27 @@
+      .supportFragmentManager
+      .findFragmentById(
+        R.id.drag_drop_test_fragment_placeholder
+      ) as DragDropTestFragment?
```
```suggestion
      ) as? DragDropTestFragment
```

I think this is what was meant (otherwise it won't cast to null when a reader might normally expect it to).

KevinGitonga replied:
Updated.

BenHenning commented on app/src/main/java/org/oppia/android/app/testing/DragDropTestFragment.kt:
```diff
@@ -0,0 +1,57 @@
+import org.oppia.android.app.story.StoryFragment
+import javax.inject.Inject
+
+/** Fragment for displaying a DragDropTestFragment. */
```
This KDoc is self-referencing--can it maybe be made a bit more clear on what the fragment is meant to do?

KevinGitonga replied:
Updated.

BenHenning commented on app/src/main/java/org/oppia/android/app/testing/DragDropTestFragment.kt:
```diff
@@ -0,0 +1,57 @@
+class DragDropTestFragment : InjectableFragment(), OnItemDragListener, OnDragEndedListener {
+
+  companion object {
+    /** Returns a new [StoryFragment] to display the story corresponding to the specified story ID. */
```
This KDoc isn't correct--please update.

KevinGitonga replied:
Noted.

BenHenning commented on app/src/main/java/org/oppia/android/app/testing/DragDropTestFragment.kt:
```diff
@@ -0,0 +1,57 @@
+    container: ViewGroup?,
+    savedInstanceState: Bundle?
+  ): View? {
+
```
Nit: Please remove blank newline (for consistency elsewhere in the codebase).

KevinGitonga replied:
Noted.

BenHenning commented on app/src/main/java/org/oppia/android/app/testing/DragDropTestFragmentPresenter.kt:
```diff
@@ -0,0 +1,79 @@
+    inflater: LayoutInflater,
+    container: ViewGroup?
+  ): View? {
+
```
Ditto here.

KevinGitonga replied:
Noted.

BenHenning commented on app/src/main/res/layout/drag_drop_test_activity.xml:
```diff
@@ -1,7 +1,7 @@
-  android:layout_height="wrap_content"
-  app:layoutManager="androidx.recyclerview.widget.LinearLayoutManager" />
+  android:layout_height="match_parent"
+  tools:context=".app.story.StoryActivity" />
```
Nit:
1. This file is missing an EOF newline.
2. This tools context seems incorrect.

KevinGitonga replied:
Updated.

BenHenning commented on app/src/main/res/layout/drag_drop_test_fragment.xml:
```diff
@@ -0,0 +1,10 @@
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    app:layoutManager="androidx.recyclerview.widget.LinearLayoutManager"/>
+</layout>
```
Please add EOF newline.

KevinGitonga replied:
updated.

BenHenning replied:
It seems to still be missing per the latest code.

KevinGitonga replied:
Updated.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/recyclerview/BindableAdapterTest.kt:
```diff
@@ -374,135 +373,6 @@ class BindableAdapterTest {
-  }
-
-  @Test
-  fun testSingleTypeAdapter_withLiveData_withLifecycleOwner_rebindsLiveDataValues() {
```
This & the multi version one below seem worth keeping as this is the behavior that should always be the case now (given that a lifecycle owner is always bound). These tests will help catch future regression for when people accidentally break the lifecycle owner part (which I think we want to make sure doesn't break given the benefits and promises of this PR).

KevinGitonga replied:
Okay, noted.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/testing/DragDropTestActivityTest.kt:
```diff
@@ -160,10 +160,11 @@ class DragDropTestActivityTest {
-        attachDragDropToActivity(activity)
-      }
+    launch(DragDropTestActivity::class.java).use {
+      // scenario ->
+      // scenario.onActivity { activity ->
+      //  attachDragDropToActivity(activity)
+      // }
```
Please update.

KevinGitonga replied:
Updated.

BenHenning reviewed (commented):
Thanks @KevinGitonga! Took another pass. I mostly have nits left, but PTAL.

Also, I think you have failing CI checks. Please try to address those before sending the PR back for review.

KevinGitonga committed:
Fix nits and updates as advised during code review.

BenHenning commented on app/src/main/java/org/oppia/android/app/recyclerview/BindableAdapter.kt:
```diff
@@ -99,7 +99,7 @@ class BindableAdapter<T : Any> internal constructor(
    * [SingleTypeBuilder] or [MultiTypeBuilder] instead.
    */
-  abstract class BaseBuilder(private val fragment: Fragment) {
+  abstract class BaseBuilder internal constructor(fragment: Fragment) {
```
Given that we no longer need to expose this class outside of the adapter, can we make both the class & its constructor private?

KevinGitonga replied:
@BenHenning changing it to private shows error "Cannot access '<init>': it is private in 'BaseBuilder'" please advice on this.

BenHenning replied:
Per our meeting, I think I suggested something along the lines of hiding the inline factory function behind a non-inline variant, but I actually don't think that's possible since we're using a reified type. This unfortunately has to be public I think.

BenHenning commented on app/src/main/java/org/oppia/android/app/recyclerview/BindableAdapter.kt:
```diff
@@ -207,7 +207,7 @@ class BindableAdapter<T : Any> internal constructor(
             setViewModel(binding, data)
             // Attach lifecycleOwner after viewModel has been attached to the view.
-            // Attaching lifecycleOwner before will cause a null pointer exception crash on [HomeFragment]
+            // Attaching lifecycleOwner before view model initialization can sometimes cause a NullPointerException because data might not be attached to the views yet.
```
This is might clearer. Two things:
- I think you can remove the 209 line comment since this covers the same thing.
- Please make sure this wraps to the 100 character line limit.

Ditto below for both.

KevinGitonga replied:
Noted.

BenHenning commented on app/src/main/java/org/oppia/android/app/testing/DragDropTestFragment.kt:
```diff
@@ -10,14 +10,18 @@ import org.oppia.android.app.fragment.FragmentComponentImpl
 import javax.inject.Inject
 
-/** Fragment for displaying a DragDropTestFragment. */
+/**
+ * Fragment for displaying a [DragDropTestFragment].
+ * Added to enable access to [BindableAdapter]
+ * which should be provided through dependency injection,
+ * this is replacing use case of [DragDropTestActivity] since adapter can not be injected to activities.
+ */
```
```suggestion
/** Test-only fragment used for verifying ``BindableAdapter`` functionality. */
```

Simplification--I don't think we need to communicate as much here.

KevinGitonga replied:
Noted

BenHenning commented on app/src/main/res/layout/drag_drop_test_activity.xml:
```diff
@@ -4,4 +4,5 @@
-  tools:context=".app.story.StoryActivity" />
\ No newline at end of file
+  tools:context=".app.testing.DragDropTestActivity">
+</FrameLayout>
```
please add EOF newline.

KevinGitonga replied:
Noted Ben

BenHenning reviewed (commented):
Thanks @KevinGitonga! Just have a few nits left--PTAL.

BenHenning commented:
@rt4914 sending this your way for codeowner review--I think you can take a look at it now.

KevinGitonga committed:
Fix nits and updates as advised during code review.

KevinGitonga committed:
Fix nits and updates as advised during code review.

rt4914 commented on app/src/main/java/org/oppia/android/app/fragment/FragmentComponentImpl.kt:
```diff
@@ -167,4 +168,5 @@ interface FragmentComponentImpl : FragmentComponent, ViewComponentBuilderInjecto
   fun inject(walkthroughFinalFragment: WalkthroughFinalFragment)
   fun inject(walkthroughTopicListFragment: WalkthroughTopicListFragment)
   fun inject(walkthroughWelcomeFragment: WalkthroughWelcomeFragment)
+  fun inject(dragDropTestFragment: DragDropTestFragment)
```
The fragments in this file are in alphabetical order. Please fix that.

rt4914 commented on app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt:
```diff
@@ -23,7 +23,8 @@ class MarkChaptersCompletedFragmentPresenter @Inject constructor(
   private val viewModelProvider: ViewModelProvider<MarkChaptersCompletedViewModel>,
-  private val modifyLessonProgressController: ModifyLessonProgressController
+  private val modifyLessonProgressController: ModifyLessonProgressController,
+  private val adapterFactory: BindableAdapter.MultiTypeBuilder.Factory
```
At some places we are using `adapterFactory` and at other places we are naming it as `multiTypeBuilderFactory`.

I suggest keeping it consistent by using `multiTypeBuilderFactory` everywhere.

KevinGitonga replied:
This is understood, Updated.

rt4914 commented on app/src/main/java/org/oppia/android/app/administratorcontrols/AdministratorControlsFragmentPresenter.kt:
```diff
@@ -31,7 +31,8 @@ import javax.inject.Inject
   private val activity: AppCompatActivity,
-  private val fragment: Fragment
+  private val fragment: Fragment,
+  private val multiTypeBuilder: BindableAdapter.MultiTypeBuilder.Factory
```
For consistency: `private val multiTypeBuilderFactory: `

KevinGitonga replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/android/app/administratorcontrols/learneranalytics/ProfileAndDeviceIdFragmentPresenter.kt:
```diff
@@ -16,8 +16,10 @@ import javax.inject.Inject
   private val fragment: Fragment,
-  private val profileListViewModelFactory: ProfileListViewModel.Factory
+  private val profileListViewModelFactory: ProfileListViewModel.Factory,
+  private val adapterFactory: BindableAdapter.MultiTypeBuilder.Factory
```
For consistency: `private val multiTypeBuilderFactory: `

KevinGitonga replied:
updated.

rt4914 commented on app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsFragmentPresenter.kt:
```diff
@@ -24,7 +24,9 @@ import javax.inject.Inject
-  private val fragment: Fragment
+  private val fragment: Fragment,
+  private val multiTypeBuilderFactory: BindableAdapter.MultiTypeBuilder.Factory
+
```
Remove this blank line

KevinGitonga replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/android/app/home/promotedlist/ComingSoonTopicsListView.kt:
```diff
@@ -45,11 +50,22 @@ class ComingSoonTopicsListView @JvmOverloads constructor(
 
   /**
    * Sets the list of coming soon topics that this view shows to the learner.
-   *
```
Undo this change.

KevinGitonga replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/android/app/home/promotedlist/PromotedStoryListView.kt:
```diff
@@ -32,10 +32,15 @@ class PromotedStoryListView @JvmOverloads constructor(
-
-    val viewComponentFactory = FragmentManager.findFragment<Fragment>(this) as ViewComponentFactory
+    val viewComponentFactory =
+      FragmentManager.findFragment<Fragment>(this) as ViewComponentFactory
```
Undo this change.

KevinGitonga replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/android/app/home/promotedlist/PromotedStoryListView.kt:
```diff
@@ -45,14 +50,35 @@ class PromotedStoryListView @JvmOverloads constructor(
    * @param newDataList the new list of stories to present
    */
   fun setPromotedStoryList(newDataList: List<PromotedStoryViewModel>?) {
+
```
Remove this blank line

KevinGitonga replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/android/app/home/promotedlist/PromotedStoryListView.kt:
```diff
@@ -45,14 +50,35 @@ class PromotedStoryListView @JvmOverloads constructor(
 
   /**
    * Sets the list of promoted stories that this view shows to the learner.
-   *
```
Undo this change

KevinGitonga replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/android/app/player/state/StatePlayerRecyclerViewAssembler.kt:
```diff
@@ -887,11 +887,13 @@ class StatePlayerRecyclerViewAssembler private constructor(
     private val resourceHandler: AppLanguageResourceHandler,
-    private val translationController: TranslationController
+    private val translationController: TranslationController,
+    private val adapterBuilderFactory: BindableAdapter.MultiTypeBuilder.Factory,
```
For consistency: `private val multiTypeBuilderFactory: `

KevinGitonga replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/android/app/player/state/StatePlayerRecyclerViewAssembler.kt:
```diff
@@ -887,11 +887,13 @@ class StatePlayerRecyclerViewAssembler private constructor(
-    private val translationController: TranslationController
+    private val translationController: TranslationController,
+    private val adapterBuilderFactory: BindableAdapter.MultiTypeBuilder.Factory,
+    private val singleAdapterFactory: BindableAdapter.SingleTypeBuilder.Factory
```
For consistency: `private val singleTypeBuilderFactory: `

KevinGitonga replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/android/app/profile/ProfileChooserFragmentPresenter.kt:
```diff
@@ -65,7 +65,8 @@ class ProfileChooserFragmentPresenter @Inject constructor(
   private val profileManagementController: ProfileManagementController,
-  private val oppiaLogger: OppiaLogger
+  private val oppiaLogger: OppiaLogger,
+  private val adapterBuilder: BindableAdapter.MultiTypeBuilder.Factory
```
For consistency: `private val multiTypeBuilderFactory: `

KevinGitonga replied:
updated.

rt4914 commented on app/src/main/java/org/oppia/android/app/profileprogress/ProfileProgressFragmentPresenter.kt:
```diff
@@ -20,12 +20,11 @@ private const val TAG_PROFILE_PICTURE_EDIT_DIALOG = "PROFILE_PICTURE_EDIT_DIALOG
-  private val fragment: Fragment
+  private val fragment: Fragment,
+  private val viewModel: ProfileProgressViewModel,
+  private val multiTypeAdapterBuilder: BindableAdapter.MultiTypeBuilder.Factory
```
For consistency: `private val multiTypeBuilderFactory: `

KevinGitonga replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/android/app/settings/profile/ProfileListFragmentPresenter.kt:
```diff
@@ -19,7 +19,8 @@ import javax.inject.Inject
   private val fragment: Fragment,
-  private val viewModelProvider: ViewModelProvider<ProfileListViewModel>
+  private val viewModelProvider: ViewModelProvider<ProfileListViewModel>,
+  private val singleTypeAdapterBuilder: BindableAdapter.SingleTypeBuilder.Factory
```
For consistency: `private val singleTypeBuilderFactory: `

KevinGitonga replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/android/app/story/StoryFragmentPresenter.kt:
```diff
@@ -49,7 +49,8 @@ class StoryFragmentPresenter @Inject constructor(
   @TopicHtmlParserEntityType private val entityType: String,
-  private val resourceHandler: AppLanguageResourceHandler
+  private val resourceHandler: AppLanguageResourceHandler,
+  private val multiTypeAdapterFactory: BindableAdapter.MultiTypeBuilder.Factory
```
For consistency: `private val multiTypeBuilderFactory: `

KevinGitonga replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/android/app/testing/BindableAdapterTestFragmentPresenter.kt:
```diff
@@ -12,6 +12,8 @@ import javax.inject.Inject
 class BindableAdapterTestFragmentPresenter @Inject constructor(
   private val fragment: Fragment,
+  private val singleTypeBuilder: BindableAdapter.SingleTypeBuilder.Factory,
+  private val multiTypeBuilder: BindableAdapter.MultiTypeBuilder.Factory,
```
Changes for consistency

KevinGitonga replied:
updated.

rt4914 commented on app/src/main/java/org/oppia/android/app/testing/BindableAdapterTestFragmentPresenter.kt:
```diff
@@ -33,6 +35,9 @@ class BindableAdapterTestFragmentPresenter @Inject constructor(
-    fun create(fragment: Fragment): BindableAdapter<BindableAdapterTestDataModel>
+    fun create(
+      singleTypeBuilder: BindableAdapter.SingleTypeBuilder.Factory,
+      multiTypeBuilder: BindableAdapter.MultiTypeBuilder.Factory
```
Ditto

KevinGitonga replied:
updated too.

rt4914 reviewed (changes requested):
There are only two types of major changes:

1. There have been some un-required changes in code, please undo them.
2. Also, please go through all the files and use the variable name consistently either `multiTypeAdapterFactory` or something else but I think consistency will be good here. Same for `singleTypeAdapterFactory` too. It can be anything but should be meaningful and consistent.

KevinGitonga committed:
Updates issues as advised on code review, fix some of the failing tests.

KevinGitonga committed:
Fix failing tests.

KevinGitonga committed:
Merge branch 'oppia:develop' into enable_dagger_graph_access_for_bindable_adapter

KevinGitonga committed:
Fix more failing tests by reverting DragDropSortInteractionView.kt and SelectionInteractionView.kt to initial state.

rt4914 commented on app/src/main/java/org/oppia/android/app/devoptions/vieweventlogs/ViewEventLogsFragmentPresenter.kt:
```diff
@@ -18,7 +18,8 @@ import javax.inject.Inject
   private val fragment: Fragment,
-  private val viewModelProvider: ViewModelProvider<ViewEventLogsViewModel>
+  private val viewModelProvider: ViewModelProvider<ViewEventLogsViewModel>,
+  private val singleTypeAdapterFactory: BindableAdapter.SingleTypeBuilder.Factory
```
For consistency: `singleTypeBuilderFactory`

KevinGitonga replied:
Updated.

BenHenning replied:
This appears addressed in the latest code. Marking as resolved since Rajat has already approved the PR.

rt4914 commented on app/src/main/java/org/oppia/android/app/devoptions/marktopicscompleted/MarkTopicsCompletedFragmentPresenter.kt:
```diff
@@ -21,7 +21,8 @@ class MarkTopicsCompletedFragmentPresenter @Inject constructor(
   private val viewModelProvider: ViewModelProvider<MarkTopicsCompletedViewModel>,
-  private val modifyLessonProgressController: ModifyLessonProgressController
+  private val modifyLessonProgressController: ModifyLessonProgressController,
+  private val singleTypeAdapterFactory: BindableAdapter.SingleTypeBuilder.Factory
```
For consistency: `singleTypeBuilderFactory`

KevinGitonga replied:
Noted.

BenHenning replied:
This appears addressed in the latest code. Marking as resolved since Rajat has already approved the PR.

rt4914 commented on app/src/main/java/org/oppia/android/app/hintsandsolution/HintsAndSolutionDialogFragmentPresenter.kt:
```diff
@@ -36,7 +36,8 @@ class HintsAndSolutionDialogFragmentPresenter @Inject constructor(
   @ExplorationHtmlParserEntityType private val entityType: String,
-  private val resourceHandler: AppLanguageResourceHandler
+  private val resourceHandler: AppLanguageResourceHandler,
+  private val multiTypeAdapterFactory: BindableAdapter.MultiTypeBuilder.Factory
```
For consistency: `multiTypeBuilderFactory`

KevinGitonga replied:
Noted.

BenHenning replied:
This appears addressed in the latest code. Marking as resolved since Rajat has already approved the PR.

rt4914 commented on app/src/main/java/org/oppia/android/app/home/HomeFragmentPresenter.kt:
```diff
@@ -41,7 +41,8 @@ class HomeFragmentPresenter @Inject constructor(
   private val resourceHandler: AppLanguageResourceHandler,
-  private val dateTimeUtil: DateTimeUtil
+  private val dateTimeUtil: DateTimeUtil,
+  private val multiTypeAdapterFactory: BindableAdapter.MultiTypeBuilder.Factory
```
For consistency: `multiTypeBuilderFactory`

KevinGitonga replied:
Noted.

BenHenning replied:
This appears addressed in the latest code. Marking as resolved since Rajat has already approved the PR.

rt4914 reviewed (approved):
LGTM,

Still one issue in multiple files, i.e.

it has been written as `singleTypeBuilderFactory ` as well as `singleTypeAdapterFactory ` similar case for `multiType` too. This can be further improved,

Not blocking PR on this trivial change.

KevinGitonga committed:
Update multiple field namings for consistency as advised by code reviewer.

BenHenning commented on app/src/main/java/org/oppia/android/app/recyclerview/BindableAdapter.kt:
```diff
@@ -334,9 +334,13 @@ class BindableAdapter<T : Any> internal constructor(
+            // NullPointerException because data might not be attached to the views yet.
+            // Skip attaching Lifecycle when StateFragment is created as this is causing NPE, yet to find fix.
+            val currentFragment = getLifecycleOwner()?.javaClass?.simpleName.toString()
+            if (currentFragment != "QuestionPlayerFragment" && currentFragment != "StateFragment") {
```
We definitely don't want to have special casing like this (plus, it won't work in alpha builds since class names are obfuscated randomly by Proguard). I think we need to figure out why the NPE is happening and fix it.

KevinGitonga replied:
Hey @BenHenning checking on this issue although i'm not able to get to its root.

BenHenning reviewed (commented):
Thanks @KevinGitonga! I think everything in the PR LGTM except the workaround for state/questions fragments. I think we need to fix those properly before the PR can be considered done.

KevinGitonga committed:
Fix failing tests and revert from previous fix, which could not be used in alpha.

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/DragDropSortInteractionView.kt:
```diff
@@ -76,6 +75,7 @@ class DragDropSortInteractionView @JvmOverloads constructor(
     //  this more efficiently and cleanly than always relying on notifying of potential changes in the adapter when the
     //  type is set (plus the type ought to be permanent).
     this.isMultipleItemsInSamePositionAllowed = isAllowed
+    singleTypeBuilderFactory = BindableAdapter.SingleTypeBuilder.Factory(Fragment())
```
Reminder to move this back to injection.

KevinGitonga replied:
Hi @BenHenning moving this back to injection does not work i believe it's due to the mentioned data race happening in the view as mentioned on comments by the previous codeowner, had mentioned this on my updates earlier. Please check and advise i reverted SelectionInteractionView to not use DI to provide BindableAdapter since due to the mentioned data race happening on attached is called very late.

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/SelectionInteractionView.kt:
```diff
@@ -65,6 +63,7 @@ class SelectionInteractionView @JvmOverloads constructor(
     //  this more efficiently and cleanly than always relying on notifying of potential changes in the adapter when the
     //  type is set (plus the type ought to be permanent).
     this.selectionItemInputType = selectionItemInputType
+    singleTypeBuilderFactory = BindableAdapter.SingleTypeBuilder.Factory(Fragment())
```
Ditto.

BenHenning replied:
Has been resolved in latest changes.

KevinGitonga replied:
@BenHenning yeah it is resolved i would say, it now works as expected with the changes. Thanks.

BenHenning committed:
Fix out-of-order binding for selection, drop/drop.

BenHenning committed:
Merge branch 'develop' into update-pr-4412

Conflicts:
	app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/devoptions/markstoriescompleted/MarkStoriesCompletedFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/devoptions/marktopicscompleted/MarkTopicsCompletedFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/hintsandsolution/HintsAndSolutionDialogFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/home/HomeFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/onboarding/OnboardingFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/options/AppLanguageFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/options/AudioLanguageFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/options/ReadingTextSizeFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/topic/lessons/TopicLessonsFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/topic/practice/TopicPracticeFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/topic/revision/TopicRevisionFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/walkthrough/topiclist/WalkthroughTopicListFragmentPresenter.kt

BenHenning committed:
Post-merge fixes.

Mainly, this fixing previous code as well as image region selection
(which whose breakage could only be detected via manually playing the
image region selection and by running the interaction's corresponding
test suite on Espresso, both of which caught different sets of issues).

The PR includes some cleanup work as well.

BenHenning committed:
Merge branch 'develop' into update-to-latest-develop

Conflicts:
	app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/devoptions/markstoriescompleted/MarkStoriesCompletedFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/devoptions/marktopicscompleted/MarkTopicsCompletedFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/hintsandsolution/HintsAndSolutionDialogFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/home/HomeFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/onboarding/OnboardingFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/options/AppLanguageFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/options/AudioLanguageFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/options/ReadingTextSizeFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/topic/lessons/TopicLessonsFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/topic/practice/TopicPracticeFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/topic/revision/TopicRevisionFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/walkthrough/topiclist/WalkthroughTopicListFragmentPresenter.kt

KevinGitonga committed:
Merge pull request #2 from BenHenning/update-to-latest-develop

Update PR #4412 to latest develop

BenHenning committed:
Merge branch 'update-to-latest-develop' into update-pr-4412

BenHenning committed:
Merge branch 'enable_dagger_graph_access_for_bindable_adapter' of https://github.com/KevinGitonga/oppia-android into update-pr-4412

BenHenning commented:
@KevinGitonga PTAL at https://github.com/KevinGitonga/oppia-android/pull/1 and added comments or merge, then assign this back to me.

Also, if you decide to merge it please also verify that it addresses your issue as expected.

KevinGitonga committed:
Merge branch 'develop' into enable_dagger_graph_access_for_bindable_adapter

KevinGitonga committed:
Merge pull request #1 from BenHenning/update-pr-4412

Update PR #4412 with various fixes [Blocked: #2]

KevinGitonga committed:
Fix lint issues after resolving merge conflicts on CI and add lessons_chapter_view.xml which is not available on my branch even after several sync's causing build errors.

KevinGitonga commented:
@BenHenning have reviewed have however noticed a few failing tests that are unrelated to this pr, that were picked up during merging of changes probably. PTAL and advice. Should i revert them or will be auto resolved when merging to develop.

BenHenning commented:
It looks like the failure has come from a bad conflict resolution in TopicLessonsFragmentPresenter after updating the branch to the latest develop changes. Some deltas were lost and need to be restored in order for that file to be correct (at the moment, it's actually regressing so the test failures are correct).

KevinGitonga committed:
Revert deleted deltas.

KevinGitonga commented:
@BenHenning reverted the deleted deltas which fixes the previously failing tests PTAL.

BenHenning commented on app/src/main/res/layout/lessons_chapter_view.xml:
```diff
@@ -0,0 +1,65 @@
+<?xml version="1.0" encoding="utf-8"?>
```
This file shouldn't be introduced--did you accidentally reintroduce it due to a recent merge? (Or maybe I did?)

KevinGitonga replied:
Yeah Ben had introduced it trying to solve merge conflicts, getting rid of it since it is not no longer required in this branch.

BenHenning reviewed (commented):
Thanks @KevinGitonga. Only had one nit that I think we need to address before the PR can be merged.

KevinGitonga committed:
Remove lessons_chapter_view.xml since it is no longer required.

BenHenning reviewed (approved):
Thanks @KevinGitonga. This LGTM!
